### PR TITLE
ta_services: add shell of keymaster TA

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Adds a shell of keymaster TA to be used to assist an OP-TEE based
keymaster implementation.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
@jbech-linaro @igoropaniuk  @vchong

**Comment by jbech-linaro** [comment]
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

File: ta_services/keymaster/include/keymaster_ta.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#define TA_KEYMASTER_UUID { 0xd36b30c7, 0x5a5f, 0x472c, \
+			    {0xaf, 0x97, 0x7f, 0x38, 0xa2, 0xed, 0xab, 0x7d } }
+
+/*
+ * Add (re-seed) caller-provided entropy to the RNG pool. Keymaster
+ * implementations need to securely mix the provided entropy into their
+ * pool, which also must contain internally-generated entropy from a
+ * hardware random number generator.
+ *
+ * in	params[0].memref: entropy input data
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply ta_services/keymaster/include/keymaster_ta.h]: I consider this an important part in the definition of the API. The one calling KEYMASTER_CMD_ADD_RNG_ENTROPY can't be expected to trace figure out which function is called in the end to know how the API is to be used, at least not in the simple cases.

[the remark this answers is not in the document]
Comment (jbech-linaro) [reply ta_services/keymaster/include/keymaster_ta.h]: Fair enough.

File: ta_services/keymaster/km.c [hunk 2]
```diff
@@ -0,0 +1,12 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
```
Comment (igoropaniuk) [inline ta_services/keymaster/km.c]: minor: SPDX format
> (Rule 1.2.1) C source: // SPDX-License-Identifier: <expression>

File: ta_services/keymaster/private_include/user_ta_header_defines.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+/* Copyright (c) 2018, Linaro Limited */
+
+#ifndef __USER_TA_HEADER_DEFINES_H
+#define __USER_TA_HEADER_DEFINES_H
+
+#include <keymaster_ta.h>
+
+#define TA_UUID				TA_KEYMASTER_UUID
+
+#define TA_FLAGS			(TA_FLAG_SINGLE_INSTANCE | \
```
Comment (igoropaniuk) [inline ta_services/keymaster/private_include/user_ta_header_defines.h]: is there any requirement that it should be `TA_FLAG_SINGLE_INSTANCE`?

Comment (jenswi-linaro) [reply ta_services/keymaster/private_include/user_ta_header_defines.h]: Let's turn that around, is there any requirement that it mustn't be single-instance? None as far as I know, single-instance is more safe though.

File: ta_services/keymaster/km.c (hunk 2 above)
Comment (jenswi-linaro) [reply ta_services/keymaster/km.c]: Good catch, let's take that in next patch now that it's merged.

**Pull request merged**